Fix inconsistency in Partial String Index with 'second' resolution

ischurov opened the pull request:
- Closes #14826
- Tests added: test_partial_string_indexes. Comprehensive testing of partial string indexes.

## Whatsnew entry
Fix inconsistency in Partial String Index with 'second' resolution. See #14826. Now if the timestamp and the index both have resolution `second`, timestamp is considered as an exact match try and not a slice. Therefore, for `Series`, scalar will be returned, for `DataFrame` `KeyError` raised.

ischurov committed:
Made this code clearer.

ischurov committed:
Fix inconsistency in Partial String Index with 'second' resolution

See #14826.

Now the following logic applies:

- If timestamp resolution is strictly less precise than index resolution, timetamp is a slice as it can (in theory) correspond to more than one elements in the index. For `Series`, `[]` should return `Series`, for `DataFrame` — `DataFrame`.
- If timestamp resolution is equal to index resolution, then timestamp is considered as an attempt to get a kind of "exact match". For `Series`, `[]` should return scalar, for `DataFrame` — try to find column with this key (if any), and most probably raise `KeyError`.
- If timestamp resolution is strictly more precise than index resolution and does not resolve to exact match, `KeyError` have to be raised in both cases.

Testsuite is updated as well.

ischurov committed:
Make flake8 happy.

jreback commented on pandas/tseries/tests/test_partial_string_indexes.py:
```diff
@@ -0,0 +1,163 @@
+import numpy as np
+
+import pandas.util.testing as tm
```
instead of creating a new files, these should go with the existing tests in test_timeseries.py

further, don't create new ways of testing, just follow along the existing.

ischurov replied:
Thanks. I'll move it to `test_timeseries.py`. Should I keep it in the separate class or unify with one of the existing? What do you mean under "new ways of testing"? Is it okay to add functions like `assert_exact` that I did?

jreback replied:
put it with the existing tests
pls following the existing testing methodologies 

iow don't create your own helpers it's not standard

jorisvandenbossche commented:
I am +1 on the idea of this PR. 
@jreback in the discussion on the issue, I interpreted your comments as opposed to this change. Is that still the case? 
The only problem is whether we should just break this behaviour, or are able to first raise a FutureWarning for this to deprecate it.

jreback commented:
@jorisvandenbossche no was fine with it. It was a fix designed for DataFrame, but broke Series (which was never tested), so for sure a bug. I don't think there is a problem with simply changing this, it would give the wrong shape back previously.

ischurov committed:
Addressing code review

- new tests moved to test_timeseries.py and refactored to avoid creating non-standard helper functions

ischurov commented:
I addressed code review comments by @jreback.

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -4911,6 +4968,63 @@ def test_partial_slice_hourly(self):
+                                                        '2012-01-01 00',
+                                                        '2012-01-01 01']),
+                       dtype=np.int64)
+
```
you are repeating lots of tests with only a slight variation

can u so it instead with a loop

ischurov replied:
I agree. Actually, it was the reason why I wrote those helper functions initially.

Okay, I united tests for all resolutions into one loop. IMO it become less readable, but reflects the logic properly.

ischurov committed:
Addressing code review: testing different combinations with the loop instead of copy-pasting of the code

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -4941,6 +4940,62 @@ def test_partial_slice_second_precision(self):
                                 lambda: s['2005-1-1 00:00:00'])
 
+    def test_partial_slicing_dataframe(self):
+        # GH14856
```
can you give a 1-2 lines about what are asserting here

ischurov replied:
Done. 67e6bab

jreback commented:
pls add a whatsnew entry (0.20.0) in other api changes. a small sub-section, showing the change.

ischurov committed:
Addressing code review: more comments added

ischurov committed:
- Whatsnew section added
- Documentation section added

ischurov committed:
- Documentation fixes

ischurov committed:
conflict PR #14856 resolved

ischurov commented:
@jreback I added whatsnew subsection and also edited documentation a little bit.

jreback commented on doc/source/timeseries.rst:
```diff
@@ -491,6 +475,51 @@ DatetimeIndex Partial String Indexing also works on DataFrames with a ``MultiInd
+
+    series_minute = pd.Series([1, 2, 3],
+                              pd.DatetimeIndex(['2011-12-31 23:59:00',
+                                                '2012-01-01 00:00:00',
```
can you add some more commentary, typically you use smaller ipython blocks, somethign like


this is seconds resolution
example here

this key is treated like this
example here

this key is treated differently


e.g.

just reads better

jreback commented on doc/source/timeseries.rst:
```diff
@@ -491,6 +475,51 @@ DatetimeIndex Partial String Indexing also works on DataFrames with a ``MultiInd
+    dft_minute['2011-12-31 23']
+
+.. warning::
+
```
clarify this a bit (as compared to the previous section). IOW this warning was in isolation before, but now you have a whole section on what works / doesn't work, so the warning needs some reworking to avoid being redundant.

ischurov committed:
- Addressing code review: documentation clarification.

ischurov committed:
- Addressing code review: added reference to new docs section in whatsnew.

ischurov committed:
resolved merge conflict in whatsnew/v0.20.0.txt

ischurov committed:
resolved merge conflict in whatsnew/v0.20.0.txt

ischurov committed:
Merge branch 'datetimeindex-slices' of https://github.com/ischurov/pandas into datetimeindex-slices

jreback commented:
thanks!

pls review the docs when built (will prob take a few hours) and issue a follow up if needed.

jreback commented:
http://pandas-docs.github.io/pandas-docs-travis/timeseries.html will show them

ischurov commented:
@jreback Thanks for the corrections. Articles are not my strong side :) I created a follow-up PR #14966 to correct a couple of typos.

Btw, this is my first PR merged to large open source project. Thanks for your guiding :)
